File: ctests/test_reduction.cpp
#include <gtest/gtest.h>
#include <torch/torch.h>

namespace
{

    torch::Device webgpu_device()
    {
        return torch::Device(torch::DeviceType::PrivateUse1);
    }

    torch::Tensor to_webgpu(const torch::Tensor &cpu)
    {
        return cpu.to(webgpu_device());
    }

} // namespace

// Mean tests
TEST(ReductionOps, Mean1D)
{
    auto a_cpu = torch::tensor({1.0f, 2.0f, 3.0f, 4.0f, 5.0f});
    auto a = to_webgpu(a_cpu);

    auto out = torch::mean(a);
    auto expected = torch::mean(a_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-4, 1e-4));
}

TEST(ReductionOps, Mean2D)
{
    auto a_cpu = torch::randn({16, 32});
    auto a = to_webgpu(a_cpu);

    auto out = torch::mean(a);
    auto expected = torch::mean(a_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-3, 1e-3));
}

TEST(ReductionOps, MeanLarge)
{
    auto a_cpu = torch::randn({64, 64});
    auto a = to_webgpu(a_cpu);

    auto out = torch::mean(a);
    auto expected = torch::mean(a_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-3, 1e-3));
}

// Sum tests
TEST(ReductionOps, Sum1D)
{
    auto a_cpu = torch::tensor({1.0f, 2.0f, 3.0f, 4.0f, 5.0f});
    auto a = to_webgpu(a_cpu);

    auto out = torch::sum(a);
    auto expected = torch::sum(a_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-4, 1e-4));
}

TEST(ReductionOps, Sum2D)
{
    auto a_cpu = torch::randn({16, 32});
    auto a = to_webgpu(a_cpu);

    auto out = torch::sum(a);
    auto expected = torch::sum(a_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-2, 1e-2));
}

TEST(ReductionOps, SumLarge)
{
    auto a_cpu = torch::randn({64, 64});
    auto a = to_webgpu(a_cpu);

    auto out = torch::sum(a);
    auto expected = torch::sum(a_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-1, 1e-1));
}

TEST(ReductionOps, SumSmallPositive)
{
    auto a_cpu = torch::ones({10, 10});
    auto a = to_webgpu(a_cpu);

    auto out = torch::sum(a);
    auto expected = torch::tensor(100.0f);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-4, 1e-4));
}

File: tests/ops/test_reduction.py
import torch
import torch_webgpu  # noqa: F401


# Mean tests
def test_mean_1d():
    a = torch.tensor([1.0, 2.0, 3.0, 4.0, 5.0], device="webgpu")
    result = torch.mean(a)
    expected = torch.tensor(3.0)
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-4, atol=1e-4)


def test_mean_2d():
    a = torch.randn(16, 32).to("webgpu")
    result = torch.mean(a)
    expected = torch.mean(a.to("cpu"))
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-3, atol=1e-3)


def test_mean_large():
    a = torch.randn(64, 64).to("webgpu")
    result = torch.mean(a)
    expected = torch.mean(a.to("cpu"))
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-3, atol=1e-3)


# Sum tests
def test_sum_1d():
    a = torch.tensor([1.0, 2.0, 3.0, 4.0, 5.0], device="webgpu")
    result = torch.sum(a)
    expected = torch.tensor(15.0)
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-4, atol=1e-4)


def test_sum_2d():
    a = torch.randn(16, 32).to("webgpu")
    result = torch.sum(a)
    expected = torch.sum(a.to("cpu"))
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-2, atol=1e-2)


def test_sum_ones():
    a = torch.ones(10, 10, device="webgpu")
    result = torch.sum(a)
    expected = torch.tensor(100.0)
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-4, atol=1e-4)
